server/src/main/java/org/elasticsearch/common/network/HandlingTimeTracker.java

```diff
@@ -63,4 +64,33 @@ public long[] getHistogram() {
         return histogram;
     }
 
+    /**
+     * Calculate the Nth percentile value
+     *
+     * @param percentile The percentile as a fraction (in [0, 1.0])
+     * @return A value greater than or equal to the specified fraction of values in the histogram
+     */
+    public long getPercentile(float percentile) {
+        assert percentile >= 0 && percentile <= 1;
+        long[] snapshot = getHistogram();
+        long totalCount = Arrays.stream(snapshot).reduce(0L, Long::sum);
+        long percentileIndex = (long) Math.ceil(totalCount * percentile);
+        for (int i = 0; i < BUCKET_COUNT; i++) {
+            percentileIndex -= snapshot[i];
+            if (percentileIndex <= 0) {
+                return getBucketUpperBounds()[i];
+            }
+        }
+        assert false : "We shouldn't ever get here";
+        return Long.MAX_VALUE;
+    }
```

Comment:
I opted to implement "inclusive" percentile because it was easier with the way the counts are stored in `HandlingTimeTracker` (the arrival of a value increments the counts of the first bucket with bound >= the value).